keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -1022,6 +1007,54 @@ is not stored, the use case is much simpler and only requires the "Subresource V
+
+These types of validations often have more complex logic and can be more difficult UX-wise to create a dedicated tag for as there are more options for representing them (tag directly on N fields, tag on one of the fields with args for the other fields, on the parent struct with args for all fields, etc.). From an analysis of current validation logic in `kubernetes/kubernetes` across native types in `pkg/apis`, a number of validation categories were identified:
+
+*   Conditional Requirement/Validation
+*   Non-Discriminated Unions
+*   Discriminated Unions
+*   List/Map Integrity
+*   Comparison
+*   Status Condition Validation
+*   Format/Value Dependencies
+    *   Rules where the validity or format of one field depends directly on the value of another field (or a value calculated from other fields). Includes: Checking if a generated string (like hostname + index) forms a valid DNS label, validating a field based on a prefix derived from another (like metadata name), or validating a field against a calculated aggregate value (like commonEncodingVersion).
+*   Transition Rules (Immutability, Ratcheting, etc.)
+*   At least "oneOf" Required
+*   Co-occurrence Requirements
+    *   Rules defining relationships where fields must appear together, be consistent if both present, or satisfy a bi-directional implication (A if and only if B).
+*   Complex/Custom Logic
+
+From this list of categories, the goal for Declarative Validation is to create dedicated tags capable of handling these categories similarly/identically to the current validation logic. The table in "Catalog of Supported Validation Rules & Associated IDL Tags" includes a number of these cross-field validation tags targeting the above categories including:
+
+*   Non-Discriminated & Discriminated Unions: `+k8s:union[Member|Discriminator]`
+*   Comparison: `+k8s:[minimum|maximum]` w/ field ref support
+*   Transition Rules - Immutability: `+k8s:immutable`.
+
+Some categories can be covered by extending and or/chaining a combination of tags. For example “Status Condition Validation” can, for identified cases (example below for CertificateSigningRequest), be handled using a combination of `+k8s:subfield` (enhanced to support targeting list entries) and `+k8s:unionMember`:
+
+```go
+type CertificateSigningRequestStatus struct {
+       // +k8s:subfield({"type":"Approved"})=+k8s:optional
+       // +k8s:subfield({"type":"Approved", "status": "true"})=+k8s:unionMember
+
+       // +k8s:subfield({"type":"Denied"})=+k8s:optional
+       // +k8s:subfield({"type":"Denied", "status": "true"})=+k8s:unionMember
+
+       // +k8s:subfield({"type":"Failed"})=+k8s:optional
+       // +k8s:subfield({"type":"Failed"})=+k8s:immutable
```

Comment:
The impl PR will go into the "immutable after create" versus "immutable once set"?

Reply:
Yes, the related impl PR will go into the "immutable after create" versus "immutable once set".